### PR TITLE
This introduces a reset functionality for object and array containers

### PR DESCRIPTION
**Pull request by lemire**
It is a private implementation that should be save to use for when we need to go at the beginning of an array or an object to go through it again.

It is not meant for end users.

### CODE REVIEW

**Comment by lemire** [comment]
@jkeiser 

I keep getting confused between `at_start` and `at_container_start`. You'd think it would be the same thing but they are very different.

Similarly `start_array` and `started_array` are very different too.

I think we need better terminology. This is not intuitive.

(Thank god it is not what end users see.)

**Comment by jkeiser** [comment]
- `at_start` means "we're at the token identifying this value" (the `[`, `"`, etc.). I imagine it's generally used when you don't know yet what type the value is. This might be renameable if we can find a name for that token (the "value token"? "first token"? "identifying token"?)
- `at_container_start`, as discussed elsewhere, can probably go away. It only applies to objects and arrays, and indicates we're in the"in between state" where we're just past the `[` but don't know whether the array/object is empty yet. The user shouldn't see us in this state unless we're non-empty.
- `at_iteration_start` means "just after start_object/start_array." If we can remove it and inline it, we should.

started_xxx could probably get a different name, though I'm not sure what offhand:

- start_array/object means "check whether the current value is an array/object, advance past the [ or {, and then start iteration."
- started_array/object means "I already know the current value is an array/object and have advanced past the [ or {, now please start iteration." It is meant for things like switch (advance()) { case '"': <process string>; case '[': started_array() }. Though we have now found uses for it in reset cases. It's possible that we could structure switch statements such that you can call start_array and the compiler will optimize the second '[' check away, though ...

File: include/simdjson/generic/ondemand/value_iterator-inl.h [hunk 1]
```diff
@@ -194,9 +194,8 @@ simdjson_warn_unused simdjson_really_inline simdjson_result<bool> value_iterator
     // this object iterator will blithely scan that object for fields.
     if (_json_iter->depth() < depth() - 1) { return OUT_OF_ORDER_ITERATION; }
 #endif
-    _json_iter->reenter_child(_start_position + 1, _depth);
```
Comment (jkeiser) [inline include/simdjson/generic/ondemand/value_iterator-inl.h:197]: Nice!

File: include/simdjson/generic/ondemand/value_iterator-inl.h [hunk 2]
```diff
@@ -628,15 +626,24 @@ inline void value_iterator::assert_at_next() const noexcept {
...
+simdjson_really_inline void value_iterator::move_at_container_start() noexcept {
+  _json_iter->_depth = _depth;
   _json_iter->token.index = _start_position + 1;
 }
 
-simdjson_really_inline void value_iterator::move_at_start() noexcept {
-    _json_iter->_depth = _depth;
-  _json_iter->token.index = _start_position;
+simdjson_really_inline bool value_iterator::reset_array() noexcept {
+  move_at_container_start();
```
Comment (jkeiser) [inline include/simdjson/generic/ondemand/value_iterator-inl.h:640]: As noted above, I'd suggest we remove move_at_container_start() and replace the calls with:

```c++
_json_iter->reenter_child(_start_position + 1, _json_iter->_depth);
```

This way, we don't have to introduce a concept for "inside the container but not knowing if it is empty yet." There is only "inside a container with a value," and "after the container."

**Review by jkeiser — APPROVED** [review]
Apart from that cosmetic fix, this feels right.

There are a lot of PRs now: did we already check in the tests you made that demonstrated the bug? (Calling count_elements on an empty array and then doing other stuff.)

**Comment by lemire** [comment]
> There are a lot of PRs now: did we already check in the tests you made that demonstrated the bug? (Calling count_elements on an empty array and then doing other stuff.)

We did.

**Comment by lemire** [comment]
I'll make the changes you suggest.

**Comment by lemire** [comment]
> `_json_iter->reenter_child(_start_position + 1, _json_iter->_depth);`

That's not equivalent to `move_at_container_start()` as I have it right now. My `move_at_container_start()` does two things...
```
  _json_iter->_depth = _depth;
  _json_iter->token.index = _start_position + 1;
```

What you propose does not reset the depth.

Should we assume that `reset_array()` or `reset_object()` is only called when `iter.depth == _json_iter->_depth`? Why would that be true? When I am inside an array or an object, and I call `reset_array()` or `reset_object()`,  how do I know that `iter.depth == _json_iter->_depth` if I do not set it or check it?

File: include/simdjson/generic/ondemand/value_iterator-inl.h (hunk 2 above)
Comment (lemire) [reply include/simdjson/generic/ondemand/value_iterator-inl.h:640]: See my comment later. I am not sure I trust this replacement to be robust.

**Comment by jkeiser** [comment]
Not at my computer right now, but I had thought reenter_child set the depth. If not, we'll have to set it explicitly, yeah.

**Comment by lemire** [comment]
@jkeiser I expect that it does set the depth but the expression you suggest will leave it unchanged because you are passing the current depth. Or that’s what I think. I could be wrong.

**Comment by jkeiser** [comment]
You are right, it looks like it needs to be depth+1. The point is not so much that we call that method though, but rather that we inline the code.

**Comment by lemire** [comment]
@jkeiser What I will do is merge this PR as-is. There is a terminology issue and that we want to reduce the number of functions, the surface area of the API... But I feel that it is better done as part of a PR dedicated to such a thing.

I think that this current PR, as is, is a step in the right direction since it makes existing code slightly simpler and slightly more robust.

There is also a need to make the internal On Demand code leaner and easier to understand... but it would be easier to appreciate it if done all at once.

**Comment by jkeiser** [comment]
Totally on board with that! There is no need to refactor adjacent concerns when solving a specific problem.

**Pull request merged**